### PR TITLE
Fix/#7862 black white screen

### PR DESCRIPTION
**Pull request by pascal-brause**
## Description
This is another try to fix the bug with the white screen. Our assumption is now that the ENF is not responding and we wait for ever for some answer.
To fix that i added a timer who checks if the ENF was activated after 3 seconds. After this time, we proceed to the Home or throw an alert to indicate that something is wrong. The fix could be to restart the phone, so for this i added a new text. @UA, please review.

If the ENF is stuck and we proceed, we will enter the the App with the status of ENF being not activated (see screenshot below). 

Only side effect I noticed is when scanning a positive test result, you cannot share this. At the point to tap the "share with others" button, an alert will occure that the ENF is not running.

To test the ENF failure behavior, comment out in `ExposureManger` the lines 229-236 and lines 280-287 in this branch.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7862

## Screenshots
App is stuck with white screen:
7862-AppStuckAtStart

App enters Home after 5 seconds of waiting and is in this status:
7862-HomeWithENFDisabled

When trying to activate the ENF, we show an alert:
7862-AlertWhenTryingToActivate

### CODE REVIEW

**Comment by pascal-brause** [comment]
refers to #2955

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 1]
```diff
@@ -220,14 +221,37 @@ final class ENAExposureManager: NSObject, ExposureManager {
...
+				timer.invalidate()
+				completion(nil)
+			} else if retries >= timeout {
+				timer.invalidate()
+				Log.error("Could not activate ENF within \(retries) seconds. Proceed with deactivated ENF")
+				completion(ExposureNotificationError.notResponding)
+			} else {
+				Log.warning("Could not activate ENF within \(retries) seconds. Retry in 1 second.")
+				retries += 1
+			}
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:254]: A solution to not check just once every second could be to make `activated` a publisher and use [timeout](https://developer.apple.com/documentation/combine/fail/timeout(_:scheduler:options:customerror:)) to wait x seconds max for it to publish. That way the execution would continue immediately.

Or an even simpler solution could be to just keep the old `completion(nil)` where it was and just use the timer to check once after x seconds if it was called and if not, call `completion(ExposureNotificationError.notResponding)`

Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift]: Good idea (last one), then i can also use GDC for the timer.

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 2]
```diff
@@ -246,17 +259,29 @@ final class ENAExposureManager: NSObject, ExposureManager {
...
 				Log.error("Failed to change ENManager.setExposureNotificationEnabled to \(status): \(error.localizedDescription)", log: .api)
 				self.handleENError(error: error, completion: completion)
 				return
 			}
+			Log.error("Successfully changed ENManager.setExposureNotificationEnabled to \(status).", log: .api)
+			changed = true
 			completion(nil)
 		}
+		
+		// Sometimes the ENF is broken. So we wait 1 seconds to ensure the changed applied. Mostly, the ENF responding instantly, so we check after 2 seconds and show then an alert.
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:276]:
```suggestion
		// Sometimes the ENF framework is broken. So we wait 1 seconds to ensure that the changed has been applied. Mostly, the ENF framework responds instantly, so we check after 2 seconds and show then an alert.
```

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 3]
```diff
@@ -246,17 +259,29 @@ final class ENAExposureManager: NSObject, ExposureManager {
 	}
 
 	private func changeEnabled(to status: Bool, completion: @escaping CompletionHandler) {
+		Log.info("Trying to change ENManager.setExposureNotificationEnabled to \(status).")
+
+		var changed = false
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:264]:
```suggestion
		var isChanged = false
```

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 4]
```diff
@@ -246,17 +259,29 @@ final class ENAExposureManager: NSObject, ExposureManager {
...
 		manager.setExposureNotificationEnabled(status) { error in
 			if let error = error {
 				Log.error("Failed to change ENManager.setExposureNotificationEnabled to \(status): \(error.localizedDescription)", log: .api)
 				self.handleENError(error: error, completion: completion)
 				return
 			}
+			Log.error("Successfully changed ENManager.setExposureNotificationEnabled to \(status).", log: .api)
+			changed = true
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:272]:
```suggestion
			isChanged = true
```

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 5]
```diff
@@ -220,14 +221,26 @@ final class ENAExposureManager: NSObject, ExposureManager {
 	/// Needs to be called before `ExposureManager.enable()`
 	func activate(completion: @escaping CompletionHandler) {
 		Log.info("Trying to activate ENManager.")
+		
+		var activated = false
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:225]:
```suggestion
		var isActivated = false
```

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 6]
```diff
@@ -220,14 +221,26 @@ final class ENAExposureManager: NSObject, ExposureManager {
...
 		manager.activate { activationError in
 			if let activationError = activationError {
 				Log.error("Failed to activate ENManager: \(activationError.localizedDescription)", log: .api)
 				self.handleENError(error: activationError, completion: completion)
 				return
 			}
-			Log.info("Activated ENManager succesfully.")
+			Log.info("Activated ENF successfully.")
 			completion(nil)
+			activated = true
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:235]:
```suggestion
			isActivated = true
```

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 7]
```diff
@@ -220,14 +221,26 @@ final class ENAExposureManager: NSObject, ExposureManager {
 	/// Needs to be called before `ExposureManager.enable()`
 	func activate(completion: @escaping CompletionHandler) {
 		Log.info("Trying to activate ENManager.")
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:223]:
```suggestion
		Log.info("Trying to activate ENF")
```

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 8]
```diff
@@ -220,14 +221,26 @@ final class ENAExposureManager: NSObject, ExposureManager {
...
 			}
-			Log.info("Activated ENManager succesfully.")
+			Log.info("Activated ENF successfully.")
 			completion(nil)
+			activated = true
+		}
+
+		// Sometimes the ENF is broken. So we check after 3 seconds if it was activated until we proceed with a deactivated ENF and log an error. Mostly, the ENF is activated instantly, so 3 seconds should be enough time to wait.
+		DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
+			if !activated {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:240]:
```suggestion
			if !isActivated {
```

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 9]
```diff
@@ -246,17 +259,29 @@ final class ENAExposureManager: NSObject, ExposureManager {
...
 				return
 			}
+			Log.error("Successfully changed ENManager.setExposureNotificationEnabled to \(status).", log: .api)
+			changed = true
 			completion(nil)
 		}
+		
+		// Sometimes the ENF is broken. So we wait 1 seconds to ensure the changed applied. Mostly, the ENF responding instantly, so we check after 2 seconds and show then an alert.
+		DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
+			if !changed {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:278]:
```suggestion
			if !isChanged {
```

**Review by naveeddotio — APPROVED** [review]
Small suggestions

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 3 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:264]: Hmm what about hasChanged?

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 4 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:272]: See above

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 6 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:235]: Proposal: isActive?

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 8 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:240]: See above

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 9 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:278]: See above

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 5 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:225]: See above

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 3 above)
Comment (naveeddotio) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift]: even better

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 6 above)
Comment (naveeddotio) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift]: even better

**Comment by jucktnich** [comment]
Are you sure you want wait 5 secs without any feedback of the app. If heard that people often restart the app from 3 secs on. Maybe show a little waiting graphic.

**Comment by pascal-brause** [comment]
> Are you sure you want to wait 5 secs without any feedback from the app. I've heard that people often restart the app from 3 secs on. Maybe show a little waiting graphic.

Thanks for your feedback. First, I decreased the time we wait to 3 seconds. Second, we can not show here a loading indicator without refactoring parts of the app start, which will take a while. And because we can not reproduce the real error, every bug fix is just an assumption so we should be minimal invasive at changing the UI for the app.

**Comment by jucktnich** [comment]
Ah yeah sometimes you should check the code and not only the description 🙈. Not showing a graphic even with only 3 secs isn't optimal, but it's really not worth the work

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 10]
```diff
@@ -219,15 +220,27 @@ final class ENAExposureManager: NSObject, ExposureManager {
 	/// Activates `ENManager`
 	/// Needs to be called before `ExposureManager.enable()`
 	func activate(completion: @escaping CompletionHandler) {
-		Log.info("Trying to activate ENManager.")
```
Comment (marcussc) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:222]: Hm the old log was more correct, wasn't it?
We activate the ENManager, which will then activate ENF right?

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 11]
```diff
@@ -219,15 +220,27 @@ final class ENAExposureManager: NSObject, ExposureManager {
...
+		var isActive = false
+		
 		manager.activate { activationError in
 			if let activationError = activationError {
 				Log.error("Failed to activate ENManager: \(activationError.localizedDescription)", log: .api)
 				self.handleENError(error: activationError, completion: completion)
 				return
 			}
-			Log.info("Activated ENManager succesfully.")
+			Log.info("Activated ENF successfully.")
```
Comment (marcussc) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:233]: Same here

**Review by marcussc — APPROVED** [review]

**Review by ChristinaTsigganos — CHANGES_REQUESTED** [review]
please take over the UA proposal for the error message

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 10 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:222]: I can revert it.

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 11 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:233]: I can revert it.

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift [hunk 12]
```diff
@@ -246,17 +259,29 @@ final class ENAExposureManager: NSObject, ExposureManager {
...
 				self.handleENError(error: error, completion: completion)
 				return
 			}
+			Log.error("Successfully changed ENManager.setExposureNotificationEnabled to \(status).", log: .api)
+			hasChanged = true
 			completion(nil)
 		}
+		
+		// Sometimes the ENF framework is broken. So we wait 1 seconds to ensure that the changed has been applied. Mostly, the ENF framework responds instantly, so we check after 2 seconds and show then an alert.
+		DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:277]: The comment say 2 seconds but the code only 1? What is the correct value herer?

**Review by KaiTeuber — APPROVED** [review]

**Comment by pascal-brause** [comment]
> please take over the UA proposal for the error message

Done.

File: src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift (hunk 12 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Models/Exposure/ExposureManager.swift:277]: the two is outdated. I'll fix that.

**Review by ChristinaTsigganos — APPROVED** [review]
UA approved

**Pull request merged**